pkg/pool-weighted/contracts/lbp/LiquidityBootstrappingPoolStorageLib.sol

```diff
@@ -0,0 +1,161 @@
+     */
+    function getGradualWeightUpdateParams(bytes32 poolState, uint256 totalTokens)
+        internal
+        pure
+        returns (
+            uint256 startTime,
+            uint256 endTime,
+            uint256[] memory startWeights,
+            uint256[] memory endWeights
+        )
+    {
+        startTime = poolState.decodeUint(_START_TIME_OFFSET, _TIMESTAMP_BIT_LENGTH);
+        endTime = poolState.decodeUint(_END_TIME_OFFSET, _TIMESTAMP_BIT_LENGTH);
+
+        startWeights = new uint256[](totalTokens);
+        endWeights = new uint256[](totalTokens);
+
+        for (uint256 i = 0; i < totalTokens; i++) {
+            startWeights[i] = poolState
+                .decodeUint(_START_WEIGHT_OFFSET + i * _START_WEIGHT_BIT_LENGTH, _START_WEIGHT_BIT_LENGTH)
+                .decompress(_START_WEIGHT_BIT_LENGTH);
+            endWeights[i] = poolState
+                .decodeUint(_END_WEIGHT_OFFSET + i * _END_WEIGHT_BIT_LENGTH, _END_WEIGHT_BIT_LENGTH)
+                .decompress(_END_WEIGHT_BIT_LENGTH);
+        }
+    }
+
+    function getWeightChangeProgress(bytes32 poolState) internal view returns (uint256) {
+        uint256 startTime = poolState.decodeUint(_START_TIME_OFFSET, _TIMESTAMP_BIT_LENGTH);
+        uint256 endTime = poolState.decodeUint(_END_TIME_OFFSET, _TIMESTAMP_BIT_LENGTH);
+
+        return GradualValueChange.calculateValueChangeProgress(startTime, endTime);
+    }
+
+    function getNormalizedWeightByIndex(
+        bytes32 poolState,
+        uint256 index,
+        uint256 pctProgress
+    ) internal pure returns (uint256) {
+        uint256 startWeight = poolState
+            .decodeUint(_START_WEIGHT_OFFSET + index * _START_WEIGHT_BIT_LENGTH, _START_WEIGHT_BIT_LENGTH)
+            .decompress(_START_WEIGHT_BIT_LENGTH);
+        uint256 endWeight = poolState
+            .decodeUint(_END_WEIGHT_OFFSET + index * _END_WEIGHT_BIT_LENGTH, _END_WEIGHT_BIT_LENGTH)
+            .decompress(_END_WEIGHT_BIT_LENGTH);
+
+        return GradualValueChange.interpolateValue(startWeight, endWeight, pctProgress);
+    }
+
+    // Setters
+
+    function setSwapEnabled(bytes32 poolState, bool swapEnabled) internal pure returns (bytes32) {
+        return poolState.insertBool(swapEnabled, _SWAP_ENABLED_OFFSET);
+    }
+
+    function setRecoveryMode(bytes32 poolState, bool enabled) internal pure returns (bytes32) {
+        return poolState.insertBool(enabled, _RECOVERY_MODE_BIT_OFFSET);
+    }
+
+    function setNormalizedWeights(
+        bytes32 poolState,
+        uint256 startTime,
+        uint256 endTime,
+        uint256[] memory newStartWeights,
+        uint256[] memory newEndWeights
+    ) internal pure returns (bytes32) {
+        poolState = poolState.insertUint(startTime, _START_TIME_OFFSET, _TIMESTAMP_BIT_LENGTH).insertUint(
+            endTime,
+            _END_TIME_OFFSET,
+            _TIMESTAMP_BIT_LENGTH
+        );
+
+        // We assume that newStartWeights.length == newEndWeights.length.
```

Comment:
```suggestion
        // This performs no checks on the validity of the weights, assuming these are done externally.
        // In particular, we assume that newStartWeights.length == newEndWeights.length, and all
        // newEndWeights are above the minimum.
```
Setting the weights is now distributed across two contracts, with the < minimum and weight sum checks done in Settings. Would like to note that somewhere for future-proofing.